postgresql/resource_postgresql_security_label.go

```diff
@@ -0,0 +1,192 @@
+		return err
+	}
+	return nil
+}
+
+func resourcePostgreSQLSecurityLabelRead(db *DBConnection, d *schema.ResourceData) error {
+	if !db.featureSupported(featureSecurityLabel) {
+		return fmt.Errorf(
+			"Security Label is not supported for this Postgres version (%s)",
+			db.version,
+		)
+	}
+	log.Printf("[WARN] PostgreSQL security label Read")
+
+	return resourcePostgreSQLSecurityLabelReadImpl(db, d)
+}
+
+func resourcePostgreSQLSecurityLabelReadImpl(db *DBConnection, d *schema.ResourceData) error {
+	objectType := d.Get(securityLabelObjectTypeAttr).(string)
+	objectName := d.Get(securityLabelObjectNameAttr).(string)
+	provider := d.Get(securityLabelProviderAttr).(string)
+
+	txn, err := startTransaction(db.client, "")
+	if err != nil {
+		return err
+	}
+	defer deferredRollback(txn)
+
+	query := "SELECT objtype, provider, label FROM pg_seclabels WHERE objtype = $1 and objname = $2 and provider = $3"
+	row := db.QueryRow(query, objectType, quoteIdentifier(objectName), quoteIdentifier(provider))
+
+	var label string
+	err = row.Scan(&objectType, &provider, &label)
+	switch {
+	case err == sql.ErrNoRows:
+		log.Printf("[WARN] PostgreSQL security label for (%s '%s') with provider %s not found", objectType, objectName, provider)
+		d.SetId("")
+		return nil
+	case err != nil:
+		return fmt.Errorf("Error reading security label: %w", err)
+	}
+
+	d.Set(securityLabelObjectTypeAttr, objectType)
+	d.Set(securityLabelObjectNameAttr, objectName)
+	d.Set(securityLabelProviderAttr, provider)
+	d.Set(securityLabelLabelAttr, label)
+	d.SetId(generateSecurityLabelID(d))
+
+	return nil
+}
+
+func resourcePostgreSQLSecurityLabelDelete(db *DBConnection, d *schema.ResourceData) error {
+	if !db.featureSupported(featureSecurityLabel) {
+		return fmt.Errorf(
+			"Security Label is not supported for this Postgres version (%s)",
+			db.version,
+		)
+	}
+	log.Printf("[WARN] PostgreSQL security label Delete")
+
+	if err := resourcePostgreSQLSecurityLabelUpdateImpl(db, d, "NULL"); err != nil {
+		return err
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func resourcePostgreSQLSecurityLabelUpdate(db *DBConnection, d *schema.ResourceData) error {
+	if !db.featureSupported(featureServer) {
+		return fmt.Errorf(
+			"Security Label is not supported for this Postgres version (%s)",
+			db.version,
+		)
+	}
+	log.Printf("[WARN] PostgreSQL security label Update")
+
+	label := d.Get(securityLabelLabelAttr).(string)
+	if err := resourcePostgreSQLSecurityLabelUpdateImpl(db, d, pq.QuoteLiteral(label)); err != nil {
+		return err
+	}
+
+	return resourcePostgreSQLSecurityLabelReadImpl(db, d)
+}
+
+func generateSecurityLabelID(d *schema.ResourceData) string {
+	return strings.Join([]string{
+		d.Get(securityLabelProviderAttr).(string),
+		d.Get(securityLabelObjectTypeAttr).(string),
+		d.Get(securityLabelObjectNameAttr).(string),
+	}, ".")
+}
+
+func quoteIdentifier(s string) string {
+	var result = s
+	re := regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*$`)
```

Comment:
which function you are referring to in `sql` lib? I probably should update name of this method, the method is actually `checkWhetherQuoteIdentifierRequired`

Reply:
Sorry, I know what you meant and which function you were referring to, this function only quote the identifier if it needs the quote. The function in `sql` library blindly quote all identifiers but the quote for the identifier that doesn't need a quote is removed from the actual value once saved, ending up with a inconsistent TF state.

Reply:
ah, right. PostgreSQL requires identifiers to be double quoted if there are any special characters in it, such as dashes and spaces, I don't think QueryRow does that, which is why we need `QuoteIdentifer` function and the test in below line.

https://github.com/cyrilgdn/terraform-provider-postgresql/blob/59f141d6e72fcf32395f4e980b01b1636e9488b9/postgresql/resource_postgresql_security_label_test.go#L204

The local `QuoteIdentifer` here though will only quote the identifier if it has such special characters to maintain consistent TF state, otherwise the SQL in the QueryRow returns empty for any identifiers that contains the above-mentioned special characters.

Reply:
I see, I indeed just tested manually  and can see that in this example the role name is stored with the dobule quotes. It's a bit weird because if you create a table with a dash, the table name will not be stored with double quotes in the pg_catalog

e.g.:
```
postgres=# create table "test-table" (id serial);
CREATE TABLE
postgres=# select tablename from pg_tables where tablename='test-table';
 tablename  
------------
 test-table
```

that's why I was expecting the `objname` to behave similarly.


Last question, why not `quoteIdentifier` everything to simplify (and avoid the need of this function)?